src/Interpreters/Cache/Metadata.cpp

```diff
@@ -293,6 +303,162 @@ void CacheMetadata::doCleanup()
+                    || file_segment->state() != FileSegment::State::PARTIALLY_DOWNLOADED)
+                    continue;
+
+                auto lock = lockKeyMetadata(file_segment->key(), KeyNotFoundPolicy::RETURN_NULL);
+                if (!lock)
+                    continue;
+
+                holder = std::make_unique<FileSegmentsHolder>(FileSegments{file_segment});
+            }
+
+            downloadImpl(holder->front());
+        }
+        catch (...)
+        {
+            if (holder)
+            {
+                const auto & file_segment = holder->front();
+                LOG_ERROR(
+                    log, "Error during background download of {}:{} ({}): {}",
+                    file_segment.key(), file_segment.offset(),
+                    file_segment.getInfoForLog(), getCurrentExceptionMessage(true));
+            }
+            else
+            {
+                tryLogCurrentException(__PRETTY_FUNCTION__);
+                chassert(false);
+            }
+        }
+    }
+}
+
+void CacheMetadata::downloadImpl(FileSegment & file_segment)
+{
+    chassert(file_segment.assertCorrectness());
+
+    if (file_segment.getOrSetDownloader() != FileSegment::getCallerId())
+        return;
+
+    LOG_TEST(
+        log, "Downloading {} bytes for file segment {}",
+        file_segment.range().size() - file_segment.getDownloadedSize(false), file_segment.getInfoForLog());
+
+    auto reader = file_segment.getRemoteFileReader();
+
+    /// If remote_fs_read_method == 'threadpool',
+    /// reader iteself bever owns/allocates the buffer.
+    std::optional<Memory<>> memory;
+    if (reader->internalBuffer().empty())
+    {
+        memory.emplace(DBMS_DEFAULT_BUFFER_SIZE);
+        reader->set(memory->data(), memory->size());
+    }
+
+    size_t offset = file_segment.getCurrentWriteOffset(false);
+    if (offset != static_cast<size_t>(reader->getPosition()))
+        reader->seek(offset, SEEK_SET);
+
+    while (!reader->eof())
+    {
+        auto size = reader->available();
+
+        if (!file_segment.reserve(size))
+        {
+            LOG_TEST(
+                log, "Failed to reserve space during background download "
+                "for {}:{} (downloaded size: {}/{})",
+                file_segment.key(), file_segment.offset(),
+                file_segment.getDownloadedSize(false), file_segment.range().size());
+            return;
+        }
+
+        try
+        {
+            file_segment.write(reader->position(), size, offset);
+            offset += size;
+            reader->position() += size;
+        }
+        catch (ErrnoException & e)
+        {
+            int code = e.getErrno();
+            if (code == /* No space left on device */28 || code == /* Quota exceeded */122)
+            {
+                LOG_INFO(log, "Insert into cache is skipped due to insufficient disk space. ({})", e.displayText());
+                return;
+            }
+            throw;
+        }
+    }
```

Comment:
shouldn't we update segment state?

Reply:
it is done in `FileSegment::write` method, there is a `try-catch` as well, which updates the file segment state and throws: https://github.com/ClickHouse/ClickHouse/blob/fcadb851c83ea35dd6ecae208c27dc3dc4fecf0b/src/Interpreters/Cache/FileSegment.cpp#L397-L398.
https://github.com/ClickHouse/ClickHouse/blob/fcadb851c83ea35dd6ecae208c27dc3dc4fecf0b/src/Interpreters/Cache/FileSegment.cpp#L533-L539
and if `FileSegment::reserve` fails - file segment state is also updates there: 
https://github.com/ClickHouse/ClickHouse/blob/fcadb851c83ea35dd6ecae208c27dc3dc4fecf0b/src/Interpreters/Cache/FileSegment.cpp#L508-L512